torch/csrc/autograd/engine.cpp

```diff
@@ -1324,6 +1360,13 @@ Engine& Engine::get_default_engine() {
   return engine_stub.load()();
 }
 
+void Engine::set_compiled_autograd(Engine::compiled_autograd_fn fn) {
+  the_compiled_autograd.store(fn);
+  TORCH_CHECK(
```

Comment:
Do the check before setting the new value.

Reply:
This would lead to a race condition and be incorrect.

Currently, thread 1 will do:
```
[A] thread_count++;
[B] the_compiled_autograd.load() 
```

Thread 2 will do:
```
[C] the_compiled_autograd.store() 
[D] thread_count.load();
```

This is correct with respect to all inter leavings of thread 1 and thread 2

However, if we swap thread 2 to do:
```
[D] thread_count.load()
[C] the_compiled_autograd.store() 
```

Then the interleaving:
`[D][A][B][C]`
would result in compiled_autograd getting enabled with a thread in eager autograd.